src/psbt.c

```diff
@@ -0,0 +1,2547 @@
+        }
+        if (!(input->final_witness = clone_witness(final_witness))) {
+            return WALLY_ENOMEM;
+        }
+    }
+    if (keypaths) {
+        if (input->keypaths) {
+            keypath_map_free(input->keypaths);
+        }
+        if (!(input->keypaths = clone_keypath_map(keypaths))) {
+            return WALLY_ENOMEM;
+        }
+    }
+    if (partial_sigs) {
+        if (input->partial_sigs) {
+            partial_sigs_map_free(input->partial_sigs);
+        }
+        if (!(input->partial_sigs = clone_partial_sigs_map(partial_sigs))) {
+            return WALLY_ENOMEM;
+        }
+    }
+    if (unknowns) {
+        if (input->unknowns) {
+            unknowns_map_free(input->unknowns);
+        }
+        if (!(input->unknowns = clone_unknowns_map(unknowns))) {
+            return WALLY_ENOMEM;
+        }
+    }
+    input->sighash_type = sighash_type;
+
+    return WALLY_OK;
+}
+
+int wally_psbt_input_free(struct wally_psbt_input *input)
+{
+    if (input) {
+        clear_and_free(input->redeem_script, input->redeem_script_len);
+        clear_and_free(input->witness_script, input->witness_script_len);
+        clear_and_free(input->final_script_sig, input->final_script_sig_len);
+        wally_tx_witness_stack_free(input->final_witness);
+        keypath_map_free(input->keypaths);
+        partial_sigs_map_free(input->partial_sigs);
+        unknowns_map_free(input->unknowns);
+    }
+    return WALLY_OK;
+}
+
+int wally_psbt_output_init_alloc(
+    unsigned char *redeem_script,
+    size_t redeem_script_len,
+    unsigned char *witness_script,
+    size_t witness_script_len,
+    struct keypath_map *keypaths,
+    struct unknowns_map *unknowns,
+    struct wally_psbt_output **output)
+{
+    struct wally_psbt_output *result;
+    int ret = WALLY_OK;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct wally_psbt_output);
+
+    ret = wally_psbt_output_update(result,
+                                   redeem_script,
+                                   redeem_script_len,
+                                   witness_script,
+                                   witness_script_len,
+                                   keypaths,
+                                   unknowns);
+
+    if (ret != WALLY_OK) {
+        clear_and_free(result, sizeof(*result));
+        *output = NULL;
+    }
+    return ret;
+}
+
+int wally_psbt_output_update(
+    struct wally_psbt_output *output,
+    unsigned char *redeem_script,
+    size_t redeem_script_len,
+    unsigned char *witness_script,
+    size_t witness_script_len,
+    struct keypath_map *keypaths,
+    struct unknowns_map *unknowns)
+{
+    if (redeem_script) {
+        if (output->redeem_script) {
+            wally_free(output->redeem_script);
+        }
+        if (clone_bytes(&output->redeem_script, redeem_script, redeem_script_len)) {
+            output->redeem_script_len = redeem_script_len;
+        } else {
+            return WALLY_ENOMEM;
```

Comment:
In this code path output->redeem_script has been freed but other data members have not and the caller cannot tell which. Maybe set the freed pointers to null?